README.md
# Bee Plugin PHP Client

![](https://raw.githubusercontent.com/kirschbaum-development/bee-plugin-php-client/master/banner.png)

[![Latest Version on Packagist](https://img.shields.io/packagist/v/kirschbaum-development/bee-plugin-php-client.svg?style=flat-square)](https://packagist.org/packages/kirschbaum-development/bee-plugin-php-client)
[![Build Status](https://travis-ci.org/kirschbaum-development/bee-plugin-php-client.svg?branch=master)](https://travis-ci.org/kirschbaum-development/bee-plugin-php-client)
[![StyleCI](https://github.styleci.io/repos/202812418/shield?branch=master)](https://github.styleci.io/repos/202812418)
[![Total Downloads](https://img.shields.io/packagist/dt/kirschbaum-development/bee-plugin-php-client.svg?style=flat-square)](https://packagist.org/packages/kirschbaum-development/bee-plugin-php-client)

PHP client to interact with Bee's plugin API. This includes the [Message Services API](https://docs.beefree.io/message-services-api-reference/) and a convenient way to use [authorization](https://docs.beefree.io/authorization-process/).

## Installation

You can install the package via composer:

```bash
composer require kirschbaum-development/bee-plugin-php-client
```

## Usage

First thing you will need is to instantiate the `Bee` base class and setup the API token. You can grab the API token by logging in into the [Developer portal](https://developers.beefree.io) and going to the "details" of your application.

``` php
use KirschbaumDevelopment\Bee\Bee;

$bee = new Bee(new \GuzzleHttp\Client);
$bee->setApiToken('your-api-token-here');
```

### Using Message Services API: HTML

This API allows you to post the JSON content generated on the Bee editor, and it will return the rendered HTML of the JSON. Reference: https://docs.beefree.io/message-services-api-reference/#html

```php
$beeEditorData = []; // this will be the JSON generated using the editor
$html = $bee->html($beeEditorData);
```

### Using Message Services API: PDF

This API allows you to post the HTML content and get an URL with a PDF created from the HTML. Reference: https://docs.beefree.io/message-services-api-reference/#pdf

```php
$pdf = $bee->pdf([
    'html' => $html, // HTML generated using the HTML API (required)
    'page_size' => 'Letter', // (optional)
    'page_orientation' => 'Landscape', // (optional)
]);

// then, you can call these methods:
$pdf->getUrl();
$pdf->getFilename();
$pdf->getPageSize();
$pdf->getPageOrientation();
$pdf->getContentType();
```

### Using Message Services API: Image

This API allows you to post the HTML content and get an URL with a PDF created from the HTML. Reference: https://docs.beefree.io/message-services-api-reference/#image

```php
$image = $bee->image([
    'html' => $html, // HTML generated using the HTML API (required)
    'width' => 500, // (required)
    'height' => 400, // (optional)
    'file_type' => 'jpg',
]);

// then, you can save the image content somewhere, some examples on how you may do this:
file_put_contents('/path/to/storage/image.jpg', $image);

// on Laravel with Storage object:
Storage::put('/path/to/storage/image.jpg', $image);
```

## Generating access token

To initialize Bee Editor you may need an access token. You can use `BeeAuth` to easily generate that. You can grab the Client ID and Secret by logging in into the [Developer portal](https://developers.beefree.io) and going to the "details" of your application.

```php
use KirschbaumDevelopment\Bee\BeeAuth;

$beeAuth = new BeeAuth(new \GuzzleHttp\Client);
$beeAuth->setClientId('your-client-id-here');
$beeAuth->setClientSecret('your-client-secret-here');
$token = $beeAuth->generateToken();

// then you can use the following methods:
$token->getAccessToken();
$token->getRefreshToken();
$token->getExpires();
```

Auth tokens are valid for 5 minutes. If you want to cache your tokens for as long as possible, you will need to pass a [PSR-16](https://www.php-fig.org/psr/psr-16/) compatible cache implementation. Laravel, Symfony and all major frameworks are already compatible with this interface. To enable the cache, you only need to call the `setCache` method passing a `Psr\SimpleCache\CacheInterface` implementation.

```php
$beeAuth->setCache($cacheImplementation);
```

## Usage with Laravel

If you use Laravel, `BeePluginServiceProvider` will be auto-registered if you use Laravel package discovery. Otherwise, you can manually register the following provider:

```php
'providers' => [
    // ...
    KirschbaumDevelopment\Bee\Laravel\BeePluginServiceProvider::class,
],
```

This will make sure that any time you ask Laravel for `KirschbaumDevelopment\Bee\Bee` or `KirschbaumDevelopment\Bee\BeeAuth` using the Laravel container (e.g. `app(Bee::class)`), Guzzle will be automatically injected and also the API token and Client ID/Secret will be injected into the classes.

The config values will be extracted from the following config: `.services.bee`. So, we'll use the following conventions:

API Token: `config('services.bee.api_token')`
Client ID: `config('services.bee.client_id')`
Client Secret: `config('services.bee.client_secret')`

If you don't want to use this config, you can always pass the values manually.

### Caching with Laravel

Laravel service provider will also automatically inject Laravel's cache implementation on `BeeAuth`, so your auth tokens will be cached for as long as possible.

### Testing with Laravel

If you are writing integration tests, but don't want your tests making requests to Bee to get credentials, we have a little helper that can help you. Just run the following code:

```php
use KirschbaumDevelopment\Bee\Laravel\BeeAuthMock;

BeeAuthMock::init();
```

With this, any time you call `BeeAuth::generateToken()` in your code you will receive the same response you would receive, but with some fake tokens and without making any HTTP requests.

## Testing

``` bash
composer test
```

## Contributing

Please see [CONTRIBUTING](CONTRIBUTING.md) for details.

### Security

If you discover any security related issues, please email luis@kirschbaumdevelopment.com or nathan@kirschbaumdevelopment.com instead of using the issue tracker.

## Credits

- [Luis Dalmolin](https://github.com/luisdalmolin)

## Sponsorship

Development of this package is sponsored by Kirschbaum Development Group, a developer driven company focused on problem solving, team building, and community. Learn more [about us](https://kirschbaumdevelopment.com) or [join us](https://careers.kirschbaumdevelopment.com)!

## License

The MIT License (MIT). Please see [License File](LICENSE.md) for more information.


CHANGELOG.md
# Changelog

All notable changes to `bee-plugin-php-client` will be documented in this file

## 1.0.0 - 201X-XX-XX

- initial release


tests/BeeTest.php
<?php

namespace KirschbaumDevelopment\Bee\Tests;

use GuzzleHttp\Client;
use GuzzleHttp\HandlerStack;
use GuzzleHttp\Psr7\Response;
use PHPUnit\Framework\TestCase;
use KirschbaumDevelopment\Bee\Bee;
use GuzzleHttp\Handler\MockHandler;
use KirschbaumDevelopment\Bee\Resources\Pdf;

/**
 * @coversDefaultClass \KirschbaumDevelopment\Bee\Bee
 */
class BeeTest extends TestCase
{
    /**
     * @covers ::html
     */
    public function testHtml()
    {
        $html = '<h1>Hello Bee.</h1>';

        $mock = new MockHandler([new Response(200, [], $html)]);
        $guzzleClient = new Client(['handler' => HandlerStack::create($mock)]);

        $beeClient = new Bee($guzzleClient);
        $this->assertEquals($html, $beeClient->html(['page' => []]));
    }

    /**
     * @covers ::pdf
     */
    public function testPdf()
    {
        $html = '<h1>Hello Bee.</h1>';
        $response = [
            'url' => 'https://www.google.com/file.pdf',
            'filename' => 'file.pdf',
            'page_size' => 'full',
            'page_orientation' => 'portrait',
            'content_type' => 'application/pdf',
        ];

        $mock = new MockHandler([new Response(200, [], json_encode($response))]);
        $guzzleClient = new Client(['handler' => HandlerStack::create($mock)]);

        $beeClient = new Bee($guzzleClient);
        $pdf = $beeClient->pdf(['html' => $html]);

        $this->assertInstanceOf(Pdf::class, $pdf);
        $this->assertEquals($response['url'], $pdf->getUrl());
        $this->assertEquals($response['filename'], $pdf->getFilename());
        $this->assertEquals($response['page_size'], $pdf->getPageSize());
        $this->assertEquals($response['page_orientation'], $pdf->getPageOrientation());
        $this->assertEquals($response['content_type'], $pdf->getContentType());
    }

    /**
     * @covers ::image
     */
    public function testImage()
    {
        $html = '<h1>Hello Bee.</h1>';
        $response = 'raw image data here';

        $mock = new MockHandler([new Response(200, [], $response)]);
        $guzzleClient = new Client(['handler' => HandlerStack::create($mock)]);

        $beeClient = new Bee($guzzleClient);
        $this->assertEquals($response, $beeClient->image([
            'html' => $html,
        ]));
    }
}


tests/BeeAuthTest.php
<?php

namespace KirschbaumDevelopment\Bee\Tests;

use GuzzleHttp\Client;
use GuzzleHttp\HandlerStack;
use GuzzleHttp\Psr7\Response;
use PHPUnit\Framework\TestCase;
use GuzzleHttp\Handler\MockHandler;
use KirschbaumDevelopment\Bee\BeeAuth;
use Cache\Adapter\PHPArray\ArrayCachePool;
use KirschbaumDevelopment\Bee\Resources\AuthorizationToken;

/**
 * @coversDefaultClass \KirschbaumDevelopment\Bee\BeeAuth
 */
class BeeAuthTest extends TestCase
{
    /**
     * @covers ::generateToken
     * @dataProvider data
     */
    public function testGenerateToken($response)
    {
        $mock = new MockHandler([new Response(200, [], json_encode($response))]);
        $guzzleClient = new Client(['handler' => HandlerStack::create($mock)]);

        $beeClient = new BeeAuth($guzzleClient);
        $beeClient->setClientId('fake-client-id');
        $beeClient->setClientSecret('fake-client-secret');
        $token = $beeClient->generateToken();

        $this->assertInstanceOf(AuthorizationToken::class, $token);
        $this->assertEquals($response['access_token'], $token->getAccessToken());
        $this->assertEquals($response['token_type'], $token->getTokenType());
        $this->assertEquals($response['expires_in'], $token->getExpiresIn());
        $this->assertEquals($response['refresh_token'], $token->getRefreshToken());
    }

    /**
     * @covers ::generateToken
     * @dataProvider data
     */
    public function testGenerateTokenUsingCache($response)
    {
        $cachePool = new ArrayCachePool;
        $mock = new MockHandler([new Response(200, [], json_encode($response))]);
        $guzzleClient = new Client(['handler' => HandlerStack::create($mock)]);

        $beeClient = new BeeAuth($guzzleClient);
        $beeClient->setCache($cachePool);
        $beeClient->setClientId('fake-client-id');
        $beeClient->setClientSecret('fake-client-secret');

        $token1 = $beeClient->generateToken();
        $token2 = $beeClient->generateToken();

        $this->assertEquals($token1, $token2);
        $this->assertInstanceOf(AuthorizationToken::class, $token1);
        $this->assertEquals($response['access_token'], $token1->getAccessToken());
        $this->assertEquals($response['token_type'], $token1->getTokenType());
        $this->assertEquals($response['expires_in'], $token1->getExpiresIn());
        $this->assertEquals($response['refresh_token'], $token1->getRefreshToken());
    }

    public function data()
    {
        return [
            [
                'response' => [
                    'access_token' => 'isufgisufdhguisdfhuisdf.oasihuifghsdif79s87f9sd8f7sd98f7sd98f7dsu89f.owTLmGhqv6ZJSikUhGeLBv-JANoXEAVvMZomivT2o-g',
                    'token_type' => 'bearer',
                    'expires_in' => 300,
                    'refresh_token' => '09a7sd98as7f98ds6f79sd8f.09a8s7da98sd7as98d7as9d8a7s9d8as6d9as86das98d67as.owTLmGhqv6ZJSikUhGeLBv-JANoXEAVvMZomivT2o-g',
                    'as:client_id' => 'da91fe57-1ca6-4dba-b557-61edb96f4149',
                    'userName' => 'Tn3KV7JYavWn',
                    'as:region' => 'eu-west-1',
                    '.issued' => 'Sat, 17 Aug 2019 00:08:50 GMT',
                    '.expires' => 'Sat, 17 Aug 2019 00:13:50 GMT',
                ],
            ],
        ];
    }
}


tests/Resources/AuthorizationTokenTest.php
<?php

namespace KirschbaumDevelopment\Bee\Tests\Resources;

use PHPUnit\Framework\TestCase;
use KirschbaumDevelopment\Bee\Resources\AuthorizationToken;

/**
 * @coversDefaultClass \KirschbaumDevelopment\Bee\Resources\AuthorizationToken
 */
class AuthorizationTokenTest extends TestCase
{
    /**
     * @covers ::toArray
     */
    public function testToArray()
    {
        $response = [
            'access_token' => 'isufgisufdhguisdfhuisdf.oasihuifghsdif79s87f9sd8f7sd98f7sd98f7dsu89f.owTLmGhqv6ZJSikUhGeLBv-JANoXEAVvMZomivT2o-g',
            'token_type' => 'bearer',
            'expires_in' => 300,
            'refresh_token' => '09a7sd98as7f98ds6f79sd8f.09a8s7da98sd7as98d7as9d8a7s9d8as6d9as86das98d67as.owTLmGhqv6ZJSikUhGeLBv-JANoXEAVvMZomivT2o-g',
        ];

        $token = new AuthorizationToken($response);
        $this->assertEquals($response, $token->toArray());
    }
}


src/Bee.php
<?php

namespace KirschbaumDevelopment\Bee;

use Exception;
use GuzzleHttp\Client as GuzzleClient;
use KirschbaumDevelopment\Bee\Resources\Pdf;

class Bee
{
    const API_BASE_URL = 'https://api.getbee.io/v1/message';

    /**
     * @var \GuzzleHttp\Client
     */
    protected $httpClient;

    /**
     * The API token.
     *
     * @var string
     */
    protected $apiToken;

    /**
     * @param \GuzzleHttp\Client $httpClient
     */
    public function __construct(GuzzleClient $httpClient)
    {
        $this->httpClient = $httpClient;
    }

    /**
     * Set the API token.
     *
     * @param string $clientId
     */
    public function setApiToken($apiToken)
    {
        $this->apiToken = $apiToken;

        return $this;
    }

    /**
     * Message API: Generate HTML from the JSON content.
     *
     * @param array $json
     * @return string
     */
    public function html(array $json)
    {
        $response = $this->httpClient->post(
            sprintf('%s/html', static::API_BASE_URL),
            [
                'headers' => $this->formatHeaders([
                    'Content-Type' => 'application/json',
                ]),
                'json' => $json,
            ]
        );

        return $response->getBody()->getContents();
    }

    /**
     * Message API: Generate PDF from the JSON content.
     *
     * @param array $payload
     * @return string
     */
    public function pdf(array $payload)
    {
        if (! isset($payload['html'])) {
            throw new Exception('"html" parameter is required to generate PDFs');
        }

        $response = $this->httpClient->post(
            sprintf('%s/pdf', static::API_BASE_URL),
            [
                'headers' => $this->formatHeaders([
                    'Content-Type' => 'application/json',
                ]),
                'json' => $payload,
            ]
        );

        return new Pdf(json_decode($response->getBody()->getContents(), true));
    }

    /**
     * Message API: Generate image from the JSON content.
     *
     * @param array $payload
     * @return string
     */
    public function image(array $payload)
    {
        $response = $this->httpClient->post(
            sprintf('%s/image', static::API_BASE_URL),
            [
                'headers' => $this->formatHeaders([
                    'Content-Type' => 'application/json',
                ]),
                'json' => $payload,
            ]
        );

        return $response->getBody();
    }

    /**
     * Return the default headers.
     *
     * @param array $headers
     * @return array
     */
    protected function formatHeaders($headers = [])
    {
        return array_merge([
            'Authorization' => sprintf('Bearer %s', $this->apiToken),
        ], $headers);
    }
}


src/BeeAuth.php
<?php

namespace KirschbaumDevelopment\Bee;

use Psr\SimpleCache\CacheInterface;
use GuzzleHttp\Client as GuzzleClient;
use KirschbaumDevelopment\Bee\Resources\AuthorizationToken;

class BeeAuth
{
    const API_AUTH_URL = 'https://auth.getbee.io/apiauth';

    /**
     * @var \GuzzleHttp\Client
     */
    protected $httpClient;

    /**
     * @var string
     */
    protected $clientId;

    /**
     * @var string
     */
    protected $clientSecret;

    /**
     * Cache PSR compatible.
     *
     * @var \Psr\SimpleCache\CacheInterface
     */
    protected $cache;

    /**
     * @param \GuzzleHttp\Client $httpClient
     */
    public function __construct(GuzzleClient $httpClient)
    {
        $this->httpClient = $httpClient;
    }

    /**
     * Set client ID.
     *
     * @param string $clientId
     */
    public function setClientId($clientId)
    {
        $this->clientId = $clientId;

        return $this;
    }

    /**
     * Set client secret.
     *
     * @param string $clientSecret
     */
    public function setClientSecret($clientSecret)
    {
        $this->clientSecret = $clientSecret;

        return $this;
    }

    /**
     * Set the cache.
     *
     * @param Psr\SimpleCache\CacheInterface $cache
     */
    public function setCache(CacheInterface $cache)
    {
        $this->cache = $cache;

        return $this;
    }

    /**
     * Return the 'Authorization' header value.
     *
     * @return string
     */
    public function generateToken()
    {
        if ($this->cache && $this->cache->has($this->getCacheKey())) {
            return new AuthorizationToken($this->cache->get($this->getCacheKey()));
        }

        $response = $this->httpClient->post(static::API_AUTH_URL, [
            'form_params' => [
                'grant_type' => 'password',
                'client_id' => $this->clientId,
                'client_secret' => $this->clientSecret,
            ],
        ]);

        $token = new AuthorizationToken(json_decode($response->getBody()->getContents(), true));

        if ($this->cache) {
            $this->cache->set($this->getCacheKey(), $token->toArray(), ($token->getExpiresIn() - 10));
        }

        return $token;
    }

    protected function getCacheKey()
    {
        return 'bee-plugin.bee-auth.token';
    }
}


src/Resources/Pdf.php
<?php

namespace KirschbaumDevelopment\Bee\Resources;

class Pdf
{
    /**
     * @var string
     */
    protected $url;

    /**
     * @var string
     */
    protected $filename;

    /**
     * @var string
     */
    protected $pageSize;

    /**
     * @var string
     */
    protected $pageOrientation;

    /**
     * @var string
     */
    protected $contentType;

    public function __construct($pdfData)
    {
        $this->url = $pdfData['url'] ?? null;
        $this->filename = $pdfData['filename'] ?? null;
        $this->pageSize = $pdfData['page_size'] ?? null;
        $this->pageOrientation = $pdfData['page_orientation'] ?? null;
        $this->contentType = $pdfData['content_type'] ?? 'application/pdf';
    }

    /**
     * @return string
     */
    public function getUrl()
    {
        return $this->url;
    }

    /**
     * @return string
     */
    public function getFilename()
    {
        return $this->filename;
    }

    /**
     * @return string
     */
    public function getPageSize()
    {
        return $this->pageSize;
    }

    /**
     * @return string
     */
    public function getPageOrientation()
    {
        return $this->pageOrientation;
    }

    /**
     * @return string
     */
    public function getContentType()
    {
        return $this->contentType;
    }

    /**
     * @return string
     */
    public function __toString()
    {
        return $this->getUrl();
    }
}


src/Resources/AuthorizationToken.php
<?php

namespace KirschbaumDevelopment\Bee\Resources;

class AuthorizationToken
{
    /**
     * @var string
     */
    protected $accessToken;

    /**
     * @var string
     */
    protected $tokenType;

    /**
     * @var string
     */
    protected $expiresIn;

    /**
     * @var string
     */
    protected $pageOrientation;

    /**
     * @var string
     */
    protected $refreshToken;

    public function __construct($tokenData)
    {
        $this->accessToken = $tokenData['access_token'] ?? null;
        $this->tokenType = $tokenData['token_type'] ?? null;
        $this->expiresIn = $tokenData['expires_in'] ?? null;
        $this->refreshToken = $tokenData['refresh_token'] ?? null;
    }

    /**
     * @return string
     */
    public function getAccessToken()
    {
        return $this->accessToken;
    }

    /**
     * @return string
     */
    public function getTokenType()
    {
        return $this->tokenType;
    }

    /**
     * @return string
     */
    public function getExpiresIn()
    {
        return $this->expiresIn;
    }

    /**
     * @return string
     */
    public function getRefreshToken()
    {
        return $this->refreshToken;
    }

    /**
     * @return array
     */
    public function toArray()
    {
        return [
            'access_token' => $this->accessToken,
            'token_type' => $this->tokenType,
            'expires_in' => $this->expiresIn,
            'refresh_token' => $this->refreshToken,
        ];
    }
}


src/Laravel/BeePluginServiceProvider.php
<?php

namespace KirschbaumDevelopment\Bee\Laravel;

use GuzzleHttp\Client;
use KirschbaumDevelopment\Bee\Bee;
use KirschbaumDevelopment\Bee\BeeAuth;
use Illuminate\Support\ServiceProvider;

class BeePluginServiceProvider extends ServiceProvider
{
    public function boot()
    {
        $this->app->bind(Bee::class, function () {
            $client = new Bee(resolve(Client::class));

            if (config('services.bee.api_token')) {
                $client->setApiToken(config('services.bee.api_token'));
            }

            return $client;
        });

        $this->app->bind(BeeAuth::class, function () {
            $client = new BeeAuth(resolve(Client::class));
            $client->setCache(app('cache.store'));

            if (config('services.bee.client_id')) {
                $client->setClientId(config('services.bee.client_id'));
            }

            if (config('services.bee.client_secret')) {
                $client->setClientSecret(config('services.bee.client_secret'));
            }

            return $client;
        });
    }
}


src/Laravel/BeeAuthMock.php
<?php

namespace KirschbaumDevelopment\Bee\Laravel;

use Mockery;
use KirschbaumDevelopment\Bee\BeeAuth;
use KirschbaumDevelopment\Bee\Resources\AuthorizationToken;

class BeeAuthMock
{
    public static function init()
    {
        $token = new AuthorizationToken([
            'access_token' => 'isufgisufdhguisdfhuisdf.oasihuifghsdif79s87f9sd8f7sd98f7sd98f7dsu89f.owTLmGhqv6ZJSikUhGeLBv-JANoXEAVvMZomivT2o-g',
            'token_type' => 'bearer',
            'expires_in' => 300,
            'refresh_token' => '09a7sd98as7f98ds6f79sd8f.09a8s7da98sd7as98d7as9d8a7s9d8as6d9as86das98d67as.owTLmGhqv6ZJSikUhGeLBv-JANoXEAVvMZomivT2o-g',
            'as:client_id' => 'da91fe57-1ca6-4dba-b557-61edb96f4149',
            'userName' => 'Tn3KV7JYavWn',
            'as:region' => 'eu-west-1',
            '.issued' => 'Sat, 17 Aug 2019 00:08:50 GMT',
            '.expires' => 'Sat, 17 Aug 2019 00:13:50 GMT',
        ]);

        $beeAuth = Mockery::mock(BeeAuth::class);
        $beeAuth->shouldReceive('generateToken')->andReturn($token);

        app()->instance(BeeAuth::class, $beeAuth);
    }
}
